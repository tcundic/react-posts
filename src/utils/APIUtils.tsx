import {apiEndpoint} from "../config";
import RequestOptions from "../interfaces/Request";
import Post from "../interfaces/Post";

const request = (options: RequestOptions, additionalHeaders?: Object): Promise<any> => {
    const headers = {
        'Content-Type': 'application/json',
        ...additionalHeaders
    };

    const defaults = {headers};
    options = Object.assign({}, defaults, options, {crossDomain: true});

    return fetch(options.url, options)
        .then(response => response.json());
};

export function getPosts(): Promise<Post[]> {
    return request({
        url: `${apiEndpoint}posts`,
        method: 'GET'
    });
}

export function getPostById(id: number): Promise<Post> {
    return request({
        url: `${apiEndpoint}posts/${id}`,
        method: 'GET'
    });
}
