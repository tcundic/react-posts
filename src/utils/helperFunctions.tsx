export const greet = (greeting: string, componentName: string): void => console.log(`${greeting} ${componentName}`);