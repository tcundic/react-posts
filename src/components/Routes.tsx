import {Redirect, Route, Switch} from "react-router-dom";
import logo2 from "../assets/logo2.svg";
import logo from "../assets/logo.svg";
import React, {useEffect} from "react";
import IRoute from "../interfaces/Route"
import {greet} from "../utils/helperFunctions";

const Routes = ({routes, greeting}: { routes: [IRoute], greeting: string}) => {

    useEffect(() => greet(greeting, Routes.name), [greeting]);

    return (
        <Switch>
            <Route exact path="/" render={() => <img src={logo2} alt={logo} className="home-page-logo"/>}/>
            {routes.map((route: IRoute) => {
                const Component = route.component;
                return <Route path={route.url} component={() => <Component greeting={greeting} />} key={route.url}/>
            }
            )}
            <Route render={() => <Redirect to="/"/>}/>
        </Switch>
    )
};

export default Routes;
