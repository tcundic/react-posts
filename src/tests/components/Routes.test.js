import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as Greet from "../../utils/helperFunctions";
import React from "react";
import Routes from "../../components/Routes";
import routes, { Component1, Component2 } from "../fixtures/routes";
import { BrowserRouter, MemoryRouter, Route } from "react-router-dom";

const greeting = "Hello from";
const name = "Routes";

test('should render Routes correctly', () => {
    const wrapper = shallow(<Routes routes={routes} greeting={greeting}/>);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(Route)).toHaveLength(4);
});

test('should print greeting', () => {
    const greetSpy = jest.spyOn(Greet, 'greet');
    const wrapper = mount(<BrowserRouter><Routes routes={routes} greeting={greeting}/></BrowserRouter>);
    expect(greetSpy).toHaveBeenLastCalledWith(greeting, name);
});

test('should change routes correctly', () => {
    let wrapper = mount(
        <MemoryRouter initialEntries={['/nonExistingRoute']}>
            <Routes routes={routes} greeting={greeting}/>
        </MemoryRouter>
    );

    expect(wrapper.find('img.home-page-logo')).toHaveLength(1);
    expect(wrapper.find(Component1)).toHaveLength(0);
    expect(wrapper.find(Component2)).toHaveLength(0);


    wrapper = mount(
        <MemoryRouter initialEntries={['/item1']}>
            <Routes routes={routes} greeting={greeting}/>
        </MemoryRouter>
    );

    expect(wrapper.find('img.home-page-logo')).toHaveLength(0);
    expect(wrapper.find(Component1)).toHaveLength(1);
    expect(wrapper.find(Component2)).toHaveLength(0);
});
