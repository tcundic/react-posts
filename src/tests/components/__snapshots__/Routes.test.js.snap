// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Routes correctly 1`] = `
<Switch>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    component={[Function]}
    key="/item1"
    path="/item1"
  />
  <Route
    component={[Function]}
    key="/item2"
    path="/item2"
  />
  <Route
    render={[Function]}
  />
</Switch>
`;
